File: benchmarks/fixtures/webapp_csharp/Errors/Exceptions.cs
namespace Webapp.Errors;

using System;

/// <summary>
/// Base class for all application exceptions.
/// </summary>
public class AppException : Exception
{
    public AppException(string message) : base(message) { }
}

public class ValidationException : AppException
{
    public ValidationException(string message) : base(message) { }
}

public class AuthenticationException : AppException
{
    public AuthenticationException(string message) : base(message) { }
}

public class TokenException : AppException
{
    public TokenException(string message) : base(message) { }
}

public class ExpiredTokenException : TokenException
{
    public ExpiredTokenException(string message) : base(message) { }
}

File: benchmarks/fixtures/webapp_cpp/errors/exceptions.cpp
#include "errors/exceptions.h"

#include <string>

namespace webapp {
namespace errors {

AppException::AppException(const std::string& message) : std::runtime_error(message) {}

ValidationError::ValidationError(const std::string& message) : AppException(message) {}

AuthenticationError::AuthenticationError(const std::string& message) : AppException(message) {}

TokenError::TokenError(const std::string& message) : AppException(message) {}

ExpiredTokenError::ExpiredTokenError(const std::string& message) : TokenError(message) {}

}  // namespace errors
}  // namespace webapp
